### DISCUSSION TITLE
How to add "local" includes when use bzlmod

### DISCUSSION THREAD

**Discussion by fausturs** [bazelbuild/bazel#20433] [Q&A] [answered]
How to add "local" includes when use bzlmod

Consider we are writing the cc_library for a library

In the [document](https://bazel.build/reference/be/c-cpp#cc_library_args) of cc_library, we can find an explanation of the includes parameter

> Subject to ["Make variable"](https://bazel.build/reference/be/make-variables) substitution. Each string is prepended with `-isystem` and added to `COPTS`. Unlike [COPTS](https://bazel.build/reference/be/c-cpp#cc_binary.copts), these flags are added for this rule and every rule that depends on it. (Note: not the rules it depends upon!) Be very careful, since this may have far-reaching effects. When in doubt, add "-I" flags to [COPTS](https://bazel.build/reference/be/c-cpp#cc_binary.copts) instead.

It seems that "-I" flags should be used when we want to add a search path that only takes effect when compiling this cc_library.

So if we use the relative path to the root of this library in the -I flag, e.g.

```
"-Isrc"
```

This option will not work when using bzlmod and this library deps by other code, because the library may layout in the sandbox like 

```
external/libname~version/src/...
```

Since most [make variables](https://bazel.build/reference/be/make-variables) only return path of files, it's hard to get the correct path of the lib.

 I have try

```
"-I$(dirname $(location src/include/a.h))/.."
"-I$$(dirname $(location src/include/a.h))/.."
"-I`dirname $(location src/include/a.h)`/.."
```

But it seems like bazel does not allow bash commands to be executed at this location

So my question is, what is the best way to handle this problem？

If there is no good solution for the time being, is it recommended to use the includes field directly? How will this be handled in the future, adding a local_includes parameter? Or add new make variables?

**Comment by meteorcloudy** [comment, marked as answer]
See https://github.com/bazelbuild/bazel/issues/15553#issuecomment-1134954324, you can use this workaround to get the canonical repo name.

**Reply by fmeum** [reply]
With the long-term solution probably being https://github.com/bazelbuild/bazel/issues/16472, not a Bzlmod-specific feature.

**Reply by fausturs** [reply]
I also think "local_includes" is a good idea
